pkg/kubelet/cm/dra/manager.go

```diff
@@ -175,27 +171,44 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 				return fmt.Errorf("NodePrepareResources failed for claim %s/%s: %s", reqClaim.Namespace, reqClaim.Name, result.Error)
 			}
 
-			claimInfo := claimInfos[types.UID(claimUID)]
-
-			// Add the CDI Devices returned by NodePrepareResources to
-			// the claimInfo object.
-			err = claimInfo.addCDIDevices(pluginName, result.GetCDIDevices())
-			if err != nil {
-				return fmt.Errorf("failed to add CDIDevices to claimInfo %+v: %+v", claimInfo, err)
+			resourceClaim := resourceClaims[types.UID(claimUID)]
+			if !m.cache.contains(resourceClaim.Name, pod.Namespace) {
+				// Grab the allocation.resourceHandles. If there are no
+				// allocation.resourceHandles, create a single resourceHandle with no
+				// content. This will trigger processing of this claim by a single
+				// kubelet plugin whose name matches resourceClaim.Status.DriverName.
+				resourceHandles := resourceClaim.Status.Allocation.ResourceHandles
+				if len(resourceHandles) == 0 {
+					resourceHandles = make([]resourceapi.ResourceHandle, 1)
+				}
+				claimInfo := newClaimInfo(
+					resourceClaim.Status.DriverName,
+					resourceClaim.Spec.ResourceClassName,
+					resourceClaim.UID,
+					resourceClaim.Name,
+					resourceClaim.Namespace,
+					sets.New(string(pod.UID)),
+					resourceHandles,
+					true,
+				)
+				if err = m.cache.add(&claimInfo); err != nil {
+					return fmt.Errorf("failed to cache claimInfo %+v: %w", claimInfo, err)
+				}
 			}
-			// mark claim as (successfully) prepared by manager, so next time we don't prepare it.
-			claimInfo.prepared = true
-
-			// TODO: We (re)add the claimInfo object to the cache and
-			// sync it to the checkpoint *after* the
-			// NodePrepareResources call has completed. This will cause
-			// issues if the kubelet gets restarted between
-			// NodePrepareResources and syncToCheckpoint. It will result
+			// NOTE: We update cached claimInfo object here.
+			// This can cause issues if the kubelet gets restarted between
+			// this point and syncToCheckpoint. It will result
 			// in not calling NodeUnprepareResources for this claim
 			// because no claimInfo will be synced back to the cache
-			// for it after the restart. We need to resolve this issue
-			// before moving to beta.
-			m.cache.add(claimInfo)
+			// for it after the restart.
+			// This can be solved by calling `m.cache.syncToCheckpoint`` here
+			// or inside `m.cache.update``
+			if err = m.cache.addCDIDevices(resourceClaim.Name, pod.Namespace, pluginName, result.GetCDIDevices()); err != nil {
+				return fmt.Errorf("claim %s: failed to update claimInfo cache: %w", resourceClaim.Name, err)
+			}
+			if err = m.cache.setPrepared(resourceClaim.Name, pod.Namespace); err != nil {
+				return fmt.Errorf("claim %s: failed to set prepared state: %w", resourceClaim.Name, err)
+			}
```

Comment:
As discussed before, this should be done outside the loop.

Reply:
This would require to collect claim names in the loop and then call `m.cache.setPrepared` for each of them. Just want to confirm you understand this.

BTW, should it be done outside of the inner loop (over claims) or outside of the outer loop (over batches)?

Reply:
I'd need to look more closely, but the point is that we can't mark the claim as prepared until all aspects of it have been prepared. Otherwise if something fails at any point within this loop, that claim will never be processed again (in plugins that haven't had a chance to see it yet).

Reply:
done outside of outer loop, PTAL